### ISSUE TITLE
`tcamsrc` *V0.11.1* hangs on `set_state(Gst.State.NULL)` reporting "Error closing udev monitoring pipe."

### ISSUE THREAD

**Issue by jraygauthier** [TheImagingSource/tiscamera#288]
Was not hanging using *V0.9.1*.

Camera is:

```bash
$ tcam-ctrl -l
Available devices:
Model           Type    Serial

DFK 33UX264     V4L2    33710150
```

Here's what happens:

```bash
$ 2>&1 TCAM_LOG=DEBUG GST_DEBUG=WARNING pytest my/camera/tests.py

39:54.410 INFO    contextlib.py                      112  TISCameraPipeline.start - begin
39:54.410 DEBUG   tiscamera_pipeline.py              281  TISCameraPipeline._reload_pipeline
39:54.410 DEBUG   tiscamera_connection.py            84   TISCameraConnection.make_gst_source_element
39:54.411 DEBUG   tiscamera_source_utils.py          51   tcam_make_source: {serial: 33710150}
12295052   <INFO> /build/source/src/LibraryHandle.cpp:72: Could not load library /nix/store/xa1i6f3vnkx1w8dp9pkz2hfh27jhsaih-tiscamera-0.11.1/lib/tcam-0/libtcam-aravis.so
12295070   <INFO> /build/source/src/LibraryHandle.cpp:73:     Reason: /nix/store/xa1i6f3vnkx1w8dp9pkz2hfh27jhsaih-tiscamera-0.11.1/lib/tcam-0/libtcam-aravis.so: cannot open shared object file: No such file or directory
12295074   <INFO> /build/source/src/BackendLoader.cpp:103: Could not load backend /nix/store/xa1i6f3vnkx1w8dp9pkz2hfh27jhsaih-tiscamera-0.11.1/lib/tcam-0/libtcam-aravis.so
12300726   <INFO> /build/source/src/v4l2/V4l2Device.cpp:424: Current framerate is 35 / 1 fps
12301001   <INFO> /build/source/src/v4l2/V4l2Device.cpp:1205: Encountered write only control.
12301219   <INFO> /build/source/src/v4l2/V4l2Device.cpp:874: Detected multiple exposure interfaces. Simplifying
12301255   <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:708: Found format: RGGB
12301281   <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:708: Found format: RG16
12301308   <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:708: Found format: YUYV
12301449   <INFO> /build/source/src/format.cpp:107: No string for fourcc 0x0
12301454   <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:708: Found format:
12301479   <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:708: Found format: BGR24
12301485   <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:255: Returning 5 formats.
12301489   <DEBUG> /build/source/src/PipelineManager.cpp:127: Received 5 formats.
12301519   <DEBUG> /build/source/src/PropertyHandler.cpp:273: Created 52 properties
12301573   <INFO> /build/source/src/gstreamer-1.0/tcamgstbase.cpp:1209: Format has empty fourcc. Ignoring
12301942   <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:1394: Changed ctrl Strobe Enable to value 1.
# ...
39:54.460 DEBUG   tiscamera_pipeline.py              221  TISCameraPipeline.src_vid_caps_str: video/x-bayer, format=rggb, width=2448, height=2048
39:54.461 DEBUG   tiscamera_pipeline.py              254  TISCameraPipeline.sink_vid_caps_str: video/x-raw, format=BGR
39:54.462 DEBUG   tiscamera_pipeline.py              268  TISCameraPipeline.videocrop: left=180, top=150, right=180, bottom=150
12311148   <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:1394: Changed ctrl Trigger Mode to value 1.
39:54.467 DEBUG   tiscamera_pipeline.py              416  TISCameraPipeline.stop: set_state(Gst.State.PLAYING)
39:54.468 INFO    contextlib.py                      119  TISCameraPipeline.start - end
12312831   <INFO> /build/source/src/ImageSource.cpp:68: Source changed to state STOPPED
12313041   <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:565: Stopped stream
12313062   <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:169: Requested format change to 'format=RGGB,width=2448,height=2048,framerate=35.000000' 42474752
39:54.469 DEBUG   mcu_acquisition_device_async.py    141  MCUAcquisitionDeviceAsync: start_acquisition async begin
39:54.472 DEBUG   mcu_acquisition_device_blocking_impl.py 530  Start acquisition (post VASP): True
39:54.472 DEBUG   mcu_acquisition_device_async.py    145  MCUAcquisitionDeviceAsync: start_acquisition async end
12315525   <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:391: Setting framerate to '35.000000' =  35 / 1
12315611   <INFO> /build/source/src/v4l2/V4l2Device.cpp:424: Current framerate is 35 / 1 fps
12315633   <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:235: Active format is: 'format=RGGB,width=2448,height=2048,framerate=35.000000'
12315666   <DEBUG> /build/source/src/PipelineManager.cpp:222: Found device fourcc 'RGGB' - 42474752
12315673   <DEBUG> /build/source/src/PipelineManager.cpp:222: Found device fourcc 'RG16' - 36314752
12315678   <DEBUG> /build/source/src/PipelineManager.cpp:222: Found device fourcc 'YUYV' - 56595559
12315684   <INFO> /build/source/src/format.cpp:107: No string for fourcc 0x0
12315689   <DEBUG> /build/source/src/PipelineManager.cpp:222: Found device fourcc '' - 0
12315694   <DEBUG> /build/source/src/PipelineManager.cpp:222: Found device fourcc 'BGR24' - 33524742
12315704   <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:169: Requested format change to 'format=RGGB,width=2448,height=2048,framerate=35.000000' 42474752
12315851   <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:391: Setting framerate to '35.000000' =  35 / 1
12316100   <INFO> /build/source/src/v4l2/V4l2Device.cpp:424: Current framerate is 35 / 1 fps
12316141   <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:235: Active format is: 'format=RGGB,width=2448,height=2048,framerate=35.000000'
12316152   <INFO> /build/source/src/ImageBuffer.cpp:38: allocating data buffer
# ...
12316207   <INFO> /build/source/src/PipelineManager.cpp:476: Pipeline creation successful.
12316212   <INFO> /build/source/src/PipelineManager.cpp:486: source ->  sink
12316217   <INFO> /build/source/src/ImageSink.cpp:51: Pipeline started playing
12316227   <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:1589: Will use 10 buffers
12316243   <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:1616: Queueing buffer(0x7feb0c070690) with length 5013504
12316516   <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:1625: Successfully queued v4l2_buffer
# ...
12318101   <INFO> /build/source/src/v4l2/V4l2Device.cpp:528: Starting stream in work thread.
12318173   <DEBUG> /build/source/src/ImageSource.cpp:57: PLAYING....
12318181   <INFO> /build/source/src/PipelineManager.cpp:83: All pipeline elements set to PLAYING.
39:54.750 DEBUG   contextlib.py                      112  TISCameraPipeline._on_frame_ready_from_gstreamer_thread - begin
39:54.750 DEBUG   tiscamera_pipeline.py              487  TISCameraPipeline._on_frame_ready_from_gstreamer_thread: sink.emit("pull-sample")
39:54.750 DEBUG   tiscamera_pipeline.py              495  TISCameraPipeline._on_frame_ready_from_gstreamer_thread: sample.get_buffer
39:54.751 DEBUG   tiscamera_pipeline.py              502  TISCameraPipeline._on_frame_ready_from_gstreamer_thread: buf.map(Gst.MapFlags.READ)
39:54.783 DEBUG   tiscamera_pipeline.py              524  TISCameraPipeline._on_frame_ready_from_gstreamer_thread: buf.unmap(mapinfo)
39:54.783 DEBUG   tiscamera_pipeline.py              528  TISCameraPipeline._on_frame_ready_from_gstreamer_thread: thread_safe_callback(frame)
39:54.783 DEBUG   contextlib.py                      119  TISCameraPipeline._on_frame_ready_from_gstreamer_thread - end

# ....

40:01.756 INFO    contextlib.py                      112  TISCameraPipeline.stop – begin

40:01.756 DEBUG   tiscamera_pipeline.py              456  TISCameraPipeline.stop: set_state(Gst.State.PAUSED)

40:01.756 DEBUG   tiscamera_pipeline.py              458  TISCameraPipeline.stop: set_state(Gst.State.NULL)
14496351   <INFO> /build/source/src/ImageSource.cpp:68: Source changed to state STOPPED
14496705   <INFO> /build/source/src/v4l2/V4l2Device.cpp:583: notifying callbacks about lost device
14496722   <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:565: Stopped stream
14496807   <INFO> /build/source/src/ImageSink.cpp:55: Pipeline stopped playing
0:00:39.250553024  9043 0x7feb0c00aca0 WARN                 tcamsrc gsttcamsrc.cpp:1053:gst_tcam_src_device_lost_callback:<tcamsrc73> error: Device lost (33710150)
0:00:39.250647722  9043 0x7feb0c00aca0 WARN                 tcamsrc gsttcamsrc.cpp:1061:gst_tcam_src_device_lost_callback:<tcamsrc73> error: Device lost
14497220   <ERROR> /build/source/src/v4l2/V4l2Device.cpp:104: Error closing udev monitoring pipe. write return '18446744073709551615'
0:00:39.250856659  9043      0x1e80630 ERROR                tcamsrc gsttcamsrc.cpp:1106:send_log_to_bus: Backend reported error: /build/source/src/v4l2/V4l2Device.cpp:104: Error closing udev monitoring pipe. write return '140728572244976'
```

More detailled log file, colorized (`*.html` in `ColorizedLogs.zip`) and not (`*.log` also in `PlainLogs.zip`) with various
level of `GST_DEBUG` but still `TCAM_LOG=DEBUG`. It also give a log for the
happy case (`*_stop_happy_case_no_hang_*`) and suffix `*_start_and_stop.*`
indicate the file containe the log for both the start and stop of the pipeline..

[20190607_tiscamerapipeline_stop_happy_case_no_hang_w_tcam_debug_w_gst_info_start_and_stop.log](https://github.com/TheImagingSource/tiscamera/files/3266456/20190607_tiscamerapipeline_stop_happy_case_no_hang_w_tcam_debug_w_gst_info_start_and_stop.log)

[20190607_tiscamerapipeline_stop_hang_w_tcam_debug_w_gst_warning_start_and_stop.log](https://github.com/TheImagingSource/tiscamera/files/3266457/20190607_tiscamerapipeline_stop_hang_w_tcam_debug_w_gst_warning_start_and_stop.log)

[20190607_tiscamerapipeline_stop_hang_w_tcam_debug_w_gst_warning.log](https://github.com/TheImagingSource/tiscamera/files/3266458/20190607_tiscamerapipeline_stop_hang_w_tcam_debug_w_gst_warning.log)

[20190607_tiscamerapipeline_stop_hang_w_tcam_debug_w_gst_info_start_and_stop.log](https://github.com/TheImagingSource/tiscamera/files/3266459/20190607_tiscamerapipeline_stop_hang_w_tcam_debug_w_gst_info_start_and_stop.log)

[20190607_tiscamerapipeline_stop_hang_w_tcam_debug_w_gst_info.log](https://github.com/TheImagingSource/tiscamera/files/3266460/20190607_tiscamerapipeline_stop_hang_w_tcam_debug_w_gst_info.log)

[20190607_tiscamerapipeline_stop_hang_w_tcam_debug_w_gst_debug.log](https://github.com/TheImagingSource/tiscamera/files/3266461/20190607_tiscamerapipeline_stop_hang_w_tcam_debug_w_gst_debug.log)

[ColorizedLogs.zip](https://github.com/TheImagingSource/tiscamera/files/3266467/ColorizedLogs.zip)

[PlainLogs.zip](https://github.com/TheImagingSource/tiscamera/files/3266473/PlainLogs.zip)

Note that in order to get these logs, I was forced to apply the following patch
in order to workaround a segfault:

[tiscamera/pull/287 - Comment out a line that segfaults the
app](https://github.com/TheImagingSource/tiscamera/pull/287)

This is most likely related to [tiscamera/issues/194 - How to properly close a
tcamsrc interface and
pipelines](https://github.com/TheImagingSource/tiscamera/issues/194)

This is a major problem for us, we need a quick fix.

Update 2019/06/13
------------------------

Note that we have to support 2 camera models: "DFM 37UX264-ML" and "DFK 33UX264".

The equivalent gstreamer pipeline is:

```bash
$ gst-launch-1.0 -v tcamsrc ! video/x-bayer, format=rggb, width=2448, height=2048 ! bayer2rgb ! videoconvert ! video/x-raw, format=BGR ! tee name=d ! queue ! fpsdisplaysink sync=false text-overlay=false video-sink=fakesink signal-fps-measurements=true  d. ! queue ! fakesink sync=false
```

however, the issue is also observable using the following native bgr pipeline
(only available on "DFK 33UX264"): 

```bash
$ gst-launch-1.0 -v tcamsrc ! video/x-raw, format=BGR, width=2448, height=2048 ! tee name=d ! queue ! fpsdisplaysink sync=false text-overlay=false video-sink=fakesink signal-fps-measurements=true  d. ! queue ! fakesink sync=false
```

The equivalent configuration is (this is our 0.9.1 version tough as some things
changed in 0.11.1):

```bash
$ ./config_camera_manually.sh 
Entered `./config_camera_manually.sh` script.
Applying manual camera configuration using `tcam-ctrl`.
tcam-ctrl -p -s "Strobe Enable=true" "33710150"
Found property!
Setting Strobe Enable to TRUE
tcam-ctrl -p -s "Strobe Polarity=true" "33710150"
Found property!
Setting Strobe Polarity to TRUE
tcam-ctrl -p -s "Strobe Exposure=true" "33710150"
Found property!
Setting Strobe Exposure to TRUE
tcam-ctrl -p -s "Strobe Duration=200" "33710150"
Found property!
tcam-ctrl -p -s "Strobe Delay=0" "33710150"
Found property!
tcam-ctrl -p -s "Trigger Polarity=RisingEdge" "33710150"
Found property!
tcam-ctrl -p -s "Trigger Delay (us)=0" "33710150"
Found property!
tcam-ctrl -p -s "Trigger Exposure Mode=ExposureActive" "33710150"
Found property!
tcam-ctrl -p -s "Trigger Mode=true" "33710150"
Found property!
Setting Trigger Mode to TRUE
tcam-ctrl -p -s "Gain Auto=false" "33710150"
Found property!
Setting Gain Auto to FALSE
tcam-ctrl -p -s "Gain=220" "33710150"
Found property!
tcam-ctrl -p -s "Exposure Auto=false" "33710150"
Found property!
Setting Exposure Auto to FALSE
tcam-ctrl -p -s "Exposure=10000" "33710150"
Found property!
tcam-ctrl -p -s "Hue=0" "33710150"
Found property!
tcam-ctrl -p -s "Saturation=64" "33710150"
Found property!
tcam-ctrl -p -s "Whitebalance=true" "33710150"
Found property!
Setting Whitebalance to TRUE
tcam-ctrl -p -s "Whitebalance Auto=false" "33710150"
Found property!
Setting Whitebalance Auto to FALSE
tcam-ctrl -p -s "Whitebalance Auto Preset=Any" "33710150"
Found property!
tcam-ctrl -p -s "Whitebalance Red=64" "33710150"
Found property!
tcam-ctrl -p -s "Whitebalance Green=64" "33710150"
Found property!
tcam-ctrl -p -s "Whitebalance Blue=255" "33710150"
Found property!
tcam-ctrl -p -s "White Balance Temperature Preset=Neutral White" "33710150"
Found property!
tcam-ctrl -p -s "Auto Functions ROI Control=false" "33710150"
Found property!
Setting Auto Functions ROI Control to FALSE
tcam-ctrl -p -s "Offset Auto Center=false" "33710150"
Found property!
Setting Offset Auto Center to FALSE
Done applying manual camera configuration using `tcam-ctrl`.
```

Indeed rollbacking to 0.9.1 is a possible workaround (and most likely what we
will do as our release deadline arrives quite soon), however it is not ideal
as we have two issues with 0.9.1:

 -  ASP00051-370: tiscamera V0.9.1: `tcam-ctrl -p` return no properties under
    linux kernel 4.19 nor under 5.0

    Workaround is currently to revert to using kernel 4.14.

 -  ASP00051-391: `tcamsrc` *V0.9.1* stuck on auto exposure even disregarding
    ext trigger on "DFM 37UX264-ML" when in `ExposureActive` mode

    On your issue tracker:
    [tiscamera/issues/290](https://github.com/TheImagingSource/tiscamera/issues/290)

    **This issue however is a dealbreaker for a workaround**. That is unless
    to probleme is fixed through a fw update of "DFM 37UX264-ML".

Note that we're running 0.11.1 on linux kernel 4.19 but we also tried on kernel
5.0 without more succes.

Note that we're using triggers to drive the camera in `ExposureActive` mode.

Note that the problem does not occur systematically but randomly after a series
of pipeline being built, played, and stopped when running our tests. I am
currently looking for a way to reproduce this in isolatio without much success
(for the time being) which is why I provide only logs for the moment (it would
not be possible to provide you with the whole codebase and its tests).

Note also that 0.9.1 also had a similar problem (when moving state from PLAYING
to PAUSE) and this looks much like this problem before we applied the following
changeset as a patch on top of the 0.9.1 release:

 -  [Add unlock function · TheImagingSource/tiscamera@50d4ea6](https://github.com/TheImagingSource/tiscamera/commit/50d4ea65bb3513f999bcc97cbd72f5eec936fa58)

Might be that a similar solution could be applied when moving state from PAUSE to NULL?

## Update 2019/07/02

The latest master branch (as of `b454a8bb06dbcb9e3d3434a03a72d3d2fa44c11e`) does
not fix the problem.

I reproduce systematically using:

 -  "DFM 37UX264-ML"
 -  `b454a8bb06dbcb9e3d3434a03a72d3d2fa44c11e` dev rule + user space
 -  Linux kernel 14.19
 -  The `reproduce_bug_51_384_hangs_on_set_state_null_nix.zip` bundle described below.

Here's the simplified log:

```bash
$ 2>&1 ./hw/camera_tools/reproduce_bug_51_384_hangs_on_set_state_null_nix.sh
INFO:__main__:init_gstreamer()
INFO:__main__:====== Iteration 0 =======
INFO:__main__:tcam_src = Gst.ElementFactory.make("tcamsrc")
INFO:__main__:tcam_src.get_device_serials()
INFO:__main__:tcam_src.set_property("serial", 25910066)
INFO:__main__:_configure_tcam_source: {use_soft_trig: True}
INFO:__main__:_make_pipeline
INFO:__main__:pipeline.set_state(Gst.State.PLAYING)
INFO:__main__:time.sleep(0.1)
INFO:__main__:------ Cycle 0 ------
INFO:__main__:time.sleep(0.08)
INFO:__main__:Trigger 0
INFO:__main__:tcam_src.set_tcam_property("Software Trigger", True)
INFO:__main__:time.sleep(0.15)
INFO:__main__:_on_new_frame_thread_safe: begin
INFO:__main__:_on_new_frame_thread_safe: new frame sent: {w: 2448, h: 2048}
INFO:__main__:_on_new_frame_thread_safe: end
INFO:__main__:time.sleep(0.003)
# ...
INFO:__main__:Trigger 5
INFO:__main__:tcam_src.set_tcam_property("Software Trigger", True)
INFO:__main__:time.sleep(0.14)
INFO:__main__:_on_new_frame_thread_safe: begin
INFO:__main__:_on_new_frame_thread_safe: new frame sent: {w: 2448, h: 2048}
INFO:__main__:_on_new_frame_thread_safe: end
INFO:__main__:time.sleep(0.003)
INFO:__main__:time.sleep(0.09)
INFO:__main__:------ Cycle 1 ------
INFO:__main__:time.sleep(0.08)
INFO:__main__:Trigger 0
INFO:__main__:tcam_src.set_tcam_property("Software Trigger", True)
INFO:__main__:time.sleep(0.15)
INFO:__main__:_on_new_frame_thread_safe: begin
INFO:__main__:_on_new_frame_thread_safe: new frame sent: {w: 2448, h: 2048}
INFO:__main__:_on_new_frame_thread_safe: end
INFO:__main__:time.sleep(0.003)
# ...
INFO:__main__:Trigger 5
INFO:__main__:tcam_src.set_tcam_property("Software Trigger", True)
INFO:__main__:time.sleep(0.14)
INFO:__main__:_on_new_frame_thread_safe: begin
INFO:__main__:_on_new_frame_thread_safe: new frame sent: {w: 2448, h: 2048}
INFO:__main__:_on_new_frame_thread_safe: end
INFO:__main__:time.sleep(0.003)
INFO:__main__:time.sleep(0.09)
INFO:__main__:time.sleep(0.1)
INFO:__main__:pipeline.set_state(Gst.State.PAUSED)
INFO:__main__:pipeline.set_state(Gst.State.NULL)
INFO:__main__:expected_frame_count: 12, actual_frame_count: 12, acceptable_frame_count: 9
INFO:__main__:====== Iteration 1 =======
INFO:__main__:tcam_src = Gst.ElementFactory.make("tcamsrc")
INFO:__main__:tcam_src.get_device_serials()
INFO:__main__:tcam_src.set_property("serial", 25910066)
INFO:__main__:_configure_tcam_source: {use_soft_trig: True}
INFO:__main__:_make_pipeline
INFO:__main__:pipeline.set_state(Gst.State.PLAYING)
INFO:__main__:time.sleep(0.1)
INFO:__main__:------ Cycle 0 ------
INFO:__main__:time.sleep(0.08)
INFO:__main__:Trigger 0
INFO:__main__:tcam_src.set_tcam_property("Software Trigger", True)
INFO:__main__:time.sleep(0.15)
INFO:__main__:_on_new_frame_thread_safe: begin
INFO:__main__:_on_new_frame_thread_safe: new frame sent: {w: 2448, h: 2048}
INFO:__main__:_on_new_frame_thread_safe: end
INFO:__main__:time.sleep(0.003)
# ...
INFO:__main__:Trigger 5
INFO:__main__:tcam_src.set_tcam_property("Software Trigger", True)
INFO:__main__:time.sleep(0.14)
INFO:__main__:_on_new_frame_thread_safe: begin
INFO:__main__:_on_new_frame_thread_safe: new frame sent: {w: 2448, h: 2048}
INFO:__main__:_on_new_frame_thread_safe: end
INFO:__main__:time.sleep(0.003)
INFO:__main__:time.sleep(0.09)
INFO:__main__:------ Cycle 1 ------
INFO:__main__:time.sleep(0.08)
INFO:__main__:Trigger 0
INFO:__main__:tcam_src.set_tcam_property("Software Trigger", True)
INFO:__main__:time.sleep(0.15)
INFO:__main__:_on_new_frame_thread_safe: begin
INFO:__main__:_on_new_frame_thread_safe: new frame sent: {w: 2448, h: 2048}
INFO:__main__:_on_new_frame_thread_safe: end
INFO:__main__:time.sleep(0.003)
# ...
INFO:__main__:Trigger 3
INFO:__main__:tcam_src.set_tcam_property("Software Trigger", True)
INFO:__main__:time.sleep(0.08)
INFO:__main__:_on_new_frame_thread_safe: begin
INFO:__main__:_on_new_frame_thread_safe: new frame sent: {w: 2448, h: 2048}
INFO:__main__:_on_new_frame_thread_safe: end
INFO:__main__:time.sleep(0.0)
INFO:__main__:time.sleep(0)
INFO:__main__:time.sleep(0.0)
INFO:__main__:pipeline.set_state(Gst.State.PAUSED)
INFO:__main__:pipeline.set_state(Gst.State.NULL)
INFO:__main__:expected_frame_count: 10, actual_frame_count: 10, acceptable_frame_count: 7
INFO:__main__:====== Iteration 2 =======
INFO:__main__:tcam_src = Gst.ElementFactory.make("tcamsrc")
INFO:__main__:tcam_src.get_device_serials()
INFO:__main__:tcam_src.set_property("serial", 25910066)
INFO:__main__:_configure_tcam_source: {use_soft_trig: True}
INFO:__main__:_make_pipeline
INFO:__main__:pipeline.set_state(Gst.State.PLAYING)
INFO:__main__:time.sleep(0.1)
INFO:__main__:------ Cycle 0 ------
INFO:__main__:time.sleep(0.08)
INFO:__main__:Trigger 0
INFO:__main__:tcam_src.set_tcam_property("Software Trigger", True)
INFO:__main__:time.sleep(0.15)
INFO:__main__:_on_new_frame_thread_safe: begin
INFO:__main__:_on_new_frame_thread_safe: new frame sent: {w: 2448, h: 2048}
INFO:__main__:_on_new_frame_thread_safe: end
INFO:__main__:time.sleep(0.003)
# ...
INFO:__main__:Trigger 5
INFO:__main__:tcam_src.set_tcam_property("Software Trigger", True)
INFO:__main__:time.sleep(0.14)
INFO:__main__:_on_new_frame_thread_safe: begin
INFO:__main__:_on_new_frame_thread_safe: new frame sent: {w: 2448, h: 2048}
INFO:__main__:_on_new_frame_thread_safe: end
INFO:__main__:time.sleep(0.003)
INFO:__main__:time.sleep(0.09)
INFO:__main__:------ Cycle 1 ------
INFO:__main__:time.sleep(0.08)
INFO:__main__:Trigger 0
INFO:__main__:tcam_src.set_tcam_property("Software Trigger", True)
INFO:__main__:time.sleep(0.15)
INFO:__main__:_on_new_frame_thread_safe: begin
INFO:__main__:_on_new_frame_thread_safe: new frame sent: {w: 2448, h: 2048}
INFO:__main__:_on_new_frame_thread_safe: end
INFO:__main__:time.sleep(0.003)
# ...
INFO:__main__:Trigger 5
INFO:__main__:tcam_src.set_tcam_property("Software Trigger", True)
INFO:__main__:time.sleep(0.14)
INFO:__main__:_on_new_frame_thread_safe: begin
INFO:__main__:_on_new_frame_thread_safe: new frame sent: {w: 2448, h: 2048}
INFO:__main__:_on_new_frame_thread_safe: end
INFO:__main__:time.sleep(0.003)
INFO:__main__:time.sleep(0.09)
INFO:__main__:time.sleep(0.1)
INFO:__main__:pipeline.set_state(Gst.State.PAUSED)
INFO:__main__:pipeline.set_state(Gst.State.NULL)
1575716    <INFO> /build/source/src/ImageSource.cpp:68: Source changed to state STOPPED
1576036    <INFO> /build/source/src/v4l2/V4l2Device.cpp:585: notifying callbacks about lost device
1576143    <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:567: Stopped stream
1576166    <INFO> /build/source/src/ImageSink.cpp:55: Pipeline stopped playing
0:00:07.743617003  7145 0x7f8ecc0068d0 WARN                 tcamsrc gsttcamsrc.cpp:1055:gst_tcam_src_device_lost_callback:<tcamsrc3> error: Device lost (25910066)
0:00:07.743909851  7145 0x7f8ecc0068d0 WARN                 tcamsrc gsttcamsrc.cpp:1063:gst_tcam_src_device_lost_callback:<tcamsrc3> error: Device lost
```

Note that it sometimes takes many iteration before the problem occurs.

Here's the full script bundle:

[reproduce_bug_51_384_hangs_on_set_state_null.zip](https://github.com/TheImagingSource/tiscamera/files/3351934/reproduce_bug_51_384_hangs_on_set_state_null.zip)

The full logs:

[Reproduce_51_384_Logs.zip](https://github.com/TheImagingSource/tiscamera/files/3351936/Reproduce_51_384_Logs.zip)

And the python code (depends on 2 other other modules in the bundle):

```py
#!/usr/bin/env python
import functools
import logging
import queue
import time
from typing import Callable, List, NamedTuple, Optional

from _gstreamer_source_utils import (
    Gst, init_gstreamer)

from _tiscamera_source_utils import (
    tcam_set_property,
    TcamSrcPropertyEffectiveValueNotAsSetError,
    _TcamPropertyValueType,
    TcamSrcPropertyDoesNotExistsError
)

LOGGER = logging.getLogger(__name__)


def setup_module() -> None:
    LOGGER.info("init_gstreamer()")
    init_gstreamer()  # Make sure gstreamer has been initialized.


def _make_tcam_src() -> Gst.Element:
    LOGGER.info("tcam_src = Gst.ElementFactory.make(\"tcamsrc\")")
    tcam_src = Gst.ElementFactory.make("tcamsrc")
    # It might be that gstreamer has not been initialized.
    assert tcam_src is not None
    LOGGER.info("tcam_src.get_device_serials()")
    connected_serials = tcam_src.get_device_serials()
    if not connected_serials:
        raise AssertionError("No connected camera!")
    first_serial = connected_serials[0]
    LOGGER.info("tcam_src.set_property(\"serial\", %s)", first_serial)
    tcam_src.set_property("serial", first_serial)

    return tcam_src


def _make_src_caps_filter(
        fps: Optional[int]
) -> Gst.Element:
    if fps is not None:
        fps_as_int = int(fps)
        framerate_str = ", framerate={}/1".format(fps_as_int)
    else:
        framerate_str = ""

    src_vid_caps_str = "{}, format={}, width={}, height={}{}".format(
        "video/x-bayer", "rggb", 2448, 2048, framerate_str)

    vid_caps = Gst.Caps.from_string(src_vid_caps_str)
    out = Gst.ElementFactory.make("capsfilter", "src_video_caps_filter")
    out.set_property("caps", vid_caps)
    return out


def _make_sink_caps_filter() -> Gst.Element:
    video_size_suffix_str = ""
    framerate_suffix_str = ""
    sink_vid_caps_str = "{}, format={}{}{}".format(
        "video/x-raw", "BGR",
        video_size_suffix_str, framerate_suffix_str)

    vid_caps = Gst.Caps.from_string(sink_vid_caps_str)
    out = Gst.ElementFactory.make("capsfilter", "sink_video_caps_filter")
    out.set_property("caps", vid_caps)
    return out


def _make_pipeline(
        source: Gst.Element,
        on_frame_ready_thread_safe: Callable[[Gst.Element], Gst.FlowReturn]
) -> Gst.Pipeline:
    LOGGER.info("_make_pipeline")

    src_vid_caps_filter = _make_src_caps_filter(fps=None)

    sink_vid_caps_filter = _make_sink_caps_filter()

    output = Gst.ElementFactory.make("appsink")
    output.set_property("emit-signals", True)
    output.set_property("async", True)
    output.connect("new-sample", on_frame_ready_thread_safe)

    pipeline = Gst.Pipeline.new()

    elements = []
    elements.append(source)
    elements.append(src_vid_caps_filter)

    bayer2rgb = Gst.ElementFactory.make("bayer2rgb")
    elements.append(bayer2rgb)

    videoconvert = Gst.ElementFactory.make("videoconvert")
    elements.append(videoconvert)

    elements.append(sink_vid_caps_filter)

    elements.append(output)

    for e in elements:
        pipeline.add(e)

    adjacent_elements_it = list(zip(elements, elements[1:]))
    for prev_e, next_e in adjacent_elements_it:
        prev_e.link(next_e)

    return pipeline


class _FrameInfo(NamedTuple):
    width: int
    height: int


def _on_new_frame_thread_safe(
        frame_queue: queue.Queue,
        sink: Gst.Element) -> Gst.FlowReturn:
    LOGGER.info("_on_new_frame_thread_safe: begin")
    try:
        sample = sink.emit("pull-sample")

        if not sample:
            LOGGER.warning(
                "_on_new_frame_thread_safe: Unexpected, could not pull sample.")
            return Gst.FlowReturn.ERROR

        buf = sample.get_buffer()
        caps = sample.get_caps()
        struct = caps.get_structure(0)

        try:
            res, mapinfo = buf.map(Gst.MapFlags.READ)

            # data = mapinfo.data
            # size = mapinfo.size

            width = struct.get_value("width")
            height = struct.get_value("height")
            # buf_frmt = struct.get_value("format")

            LOGGER.info(
                "_on_new_frame_thread_safe: new frame sent: "
                "{w: %s, h: %s}", width, height)
            frame_queue.put(_FrameInfo(
                width=width,
                height=height,
            ))

        finally:
            buf.unmap(mapinfo)

    except BaseException as e:
        LOGGER.exception(e)
        return Gst.FlowReturn.ERROR
    finally:
        LOGGER.info("_on_new_frame_thread_safe: end")

    return Gst.FlowReturn.OK


class _Configurator:
    def _set_property(
            self,
            source: Gst.Element,
            name: str, value: _TcamPropertyValueType) -> None:
        try:
            tcam_set_property(source, name, value)
        except TcamSrcPropertyDoesNotExistsError as e:
            if e.property_name in {"Hue", "Saturation"}:
                # Ok, "DFM" model does not support these.
                LOGGER.warning(str(e))
            else:
                raise  # re-raise
        except TcamSrcPropertyEffectiveValueNotAsSetError as e:
            if e.property_name == "Trigger Delay (us)" and e.actual == 31:
                # Ok, this is known and has no visible impact.
                pass
            else:
                raise  # re-raise


def _configure_tcam_source(
        tcam_src: Gst.Element, use_soft_trig: bool) -> None:
    LOGGER.info("_configure_tcam_source: {use_soft_trig: %s}", use_soft_trig)
    config = _Configurator()
    config._set_property(tcam_src, "Strobe Enable", True)
    config._set_property(tcam_src, "Strobe Polarity", True)
    config._set_property(tcam_src, "Strobe Exposure", True)
    config._set_property(tcam_src, "Strobe Duration", 200)
    config._set_property(tcam_src, "Strobe Delay", 0)
    config._set_property(tcam_src, "Trigger Polarity", "Rising Edge")
    config._set_property(tcam_src, "Trigger Delay (us)", 0)
    config._set_property(tcam_src, "Trigger Exposure Mode", "Frame Start")
    config._set_property(tcam_src, "Gain Auto", False)
    config._set_property(tcam_src, "Gain", 250)
    config._set_property(tcam_src, "Exposure Auto", False)
    config._set_property(tcam_src, "Exposure Time (us)", 333)
    config._set_property(tcam_src, "Exposure Auto Reference", 128)
    config._set_property(tcam_src, "Exposure Auto Upper Limit Auto", True)
    config._set_property(tcam_src, "Exposure Auto Lower Limit", 333)
    config._set_property(tcam_src, "Exposure Auto Upper Limit", 333)
    config._set_property(tcam_src, "Reverse X", False)
    config._set_property(tcam_src, "Reverse Y", False)
    config._set_property(tcam_src, "Hue", 0)
    config._set_property(tcam_src, "Saturation", 64)
    config._set_property(tcam_src, "Brightness", 240)
    config._set_property(tcam_src, "Whitebalance", True)
    config._set_property(tcam_src, "Whitebalance Auto", False)
    config._set_property(tcam_src, "White Balance Temperature Preset", "Neutral White")
    config._set_property(tcam_src, "Whitebalance Red", 81)
    config._set_property(tcam_src, "Whitebalance Green", 64)
    config._set_property(tcam_src, "Whitebalance Blue", 238)
    config._set_property(tcam_src, "Override Scanning Mode", 1)
    config._set_property(tcam_src, "Auto Functions ROI Control", False)
    config._set_property(tcam_src, "Offset Auto Center", False)
    config._set_property(tcam_src, "Highlight Reduction", False)
    config._set_property(tcam_src, "Trigger Mode", True)


class _TriggerConfig(NamedTuple):
    exposure_s: float
    off_delay_s: float


class _CycleConfig(NamedTuple):
    init_delay_s: float
    triggers: List[_TriggerConfig]
    final_delay_s: float

    def get_trigger_count(self) -> int:
        return len(self.triggers)


class _Iteration(NamedTuple):
    init_delay_s: float
    cycles: List[_CycleConfig]
    final_delay_s: float

    def get_cycle_count(self) -> int:
        return len(self.cycles)

    def get_expected_frame_count(self) -> int:
        return functools.reduce(
            lambda acc, c: acc + c.get_trigger_count(), self.cycles, 0)


# Reproducing ASP00051-384.
def _reproduce_bug() -> None:
    # Note that this seems to assert only when running on top of kernel 4.14
    # with 0.11.1 and not on 4.19.
    # 0.9.1 has no trouble with this.

    default_cycle = _CycleConfig(
        init_delay_s=0.08,
        triggers=([_TriggerConfig(0.150, 0.003)] * 5 + [_TriggerConfig(0.140, 0.003)]),
        final_delay_s=0.09
    )

    default_iteration = _Iteration(
        init_delay_s=0.1,
        cycles=[default_cycle] * 2,
        final_delay_s=0.1
    )

    interrupted_cycle = _CycleConfig(
        init_delay_s=0.08,
        triggers=([_TriggerConfig(0.150, 0.003)] * 3 + [_TriggerConfig(0.08, 0.0)]),
        final_delay_s=0
    )

    interrupted_iteration = _Iteration(
        init_delay_s=0.1,
        cycles=[default_cycle, interrupted_cycle],
        final_delay_s=0.0
    )

    # NOTE: It sometimes takes a lot of iterations before I successfully
    # reproduce the bug. One might want to increase this.
    iterations = [default_iteration, interrupted_iteration] * 50

    for iter_idx, iteration in enumerate(iterations):
        LOGGER.info("====== Iteration %s =======", iter_idx)

        frame_queue: queue.Queue[_FrameInfo] = queue.Queue()

        def on_new_frame_thread_safe(sink: Gst.Element) -> Gst.FlowReturn:
            return _on_new_frame_thread_safe(frame_queue, sink)

        tcam_src = _make_tcam_src()

        use_soft_trig = True
        _configure_tcam_source(tcam_src, use_soft_trig=use_soft_trig)

        pipeline = _make_pipeline(tcam_src, on_new_frame_thread_safe)

        LOGGER.info("pipeline.set_state(Gst.State.PLAYING)")
        pipeline.set_state(Gst.State.PLAYING)

        LOGGER.info("time.sleep(%s)", iteration.init_delay_s)
        time.sleep(iteration.init_delay_s)

        for c_idx, c in enumerate(iteration.cycles):
            LOGGER.info("------ Cycle %s ------", c_idx)
            LOGGER.info("time.sleep(%s)", c.init_delay_s)
            time.sleep(c.init_delay_s)

            for t_idx, t in enumerate(c.triggers):
                LOGGER.info("Trigger %s", t_idx)
                LOGGER.info("tcam_src.set_tcam_property(\"Software Trigger\", True)")
                tcam_src.set_tcam_property("Software Trigger", True)
                LOGGER.info("time.sleep(%s)", t.exposure_s)
                time.sleep(t.exposure_s)
                LOGGER.info("time.sleep(%s)", t.off_delay_s)
                time.sleep(t.off_delay_s)

            LOGGER.info("time.sleep(%s)", c.final_delay_s)
            time.sleep(c.final_delay_s)

        LOGGER.info("time.sleep(%s)", iteration.final_delay_s)
        time.sleep(iteration.final_delay_s)

        LOGGER.info("pipeline.set_state(Gst.State.PAUSED)")
        pipeline.set_state(Gst.State.PAUSED)
        LOGGER.info("pipeline.set_state(Gst.State.NULL)")
        pipeline.set_state(Gst.State.NULL)
        del pipeline

        # TODO: It appears that we do not cleanup to source when using the pipeline.
        # Cleanup.
        # LOGGER.info("tcam_src.set_state(Gst.State.NULL)")
        # tcam_src.set_state(Gst.State.NULL)
        #
        # NOTE: This is essential to reproduce the segfault as it forces
        # the src to be gced.
        del tcam_src

        frames = []
        try:
            while True:
                frames.append(frame_queue.get(block=False))
        except queue.Empty:
            pass

        expected_frame_count = iteration.get_expected_frame_count()
        actual_frame_count = len(frames)
        # ASP00051-389: Take into account the broken initial frames.
        acceptable_frame_count = int(expected_frame_count * 3.0 / 4.0)

        LOGGER.info(
            "expected_frame_count: %s, actual_frame_count: %s, acceptable_frame_count: %s",
            expected_frame_count, actual_frame_count, acceptable_frame_count)
        assert actual_frame_count >= acceptable_frame_count
        assert actual_frame_count <= expected_frame_count


def main() -> None:
    logging.basicConfig(level="INFO")

    setup_module()
    _reproduce_bug()


main()
```

**Comment by TIS-Stefan** [comment]
Hello

Since it is already 5pm here and Monday is a holiday, I suggest to go back to the working version.
I suppose, the pipeline you use is somewhere in the zip files or links.

Stefan

**Comment by TIS-Stefan** [comment]
Hello

"closing udev monitoring pipe" was an error indeed. It is fixed in the master branch now.

For the other error messages we would like to know, 
* which hardware do you use?
* How is your software built. Or does the same happen with   `gst-launch-1.0 tcambin ! videoconvert  ! ximagesink`

Thank you in advance!

Stefan

**Comment by jraygauthier** [comment]
> How is your software built. Or does the same happen with gst-launch-1.0 tcambin ! videoconvert ! ximagesink

Some updates in the description under "Update 2019/06/13". We did not attempt to reproduce the issue with gst-launch.

I should attempt to create a simplified python app that reproduce this at some point.

> which hardware do you use?

I am reproducing this on a pretty standard dev machine:

```bash
$ lscpu 
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
Address sizes:       39 bits physical, 48 bits virtual
CPU(s):              4
On-line CPU(s) list: 0-3
Thread(s) per core:  1
Core(s) per socket:  4
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               158
Model name:          Intel(R) Core(TM) i5-7600K CPU @ 3.80GHz
Stepping:            9
CPU MHz:             4000.079
CPU max MHz:         4200.0000
CPU min MHz:         800.0000
BogoMIPS:            7584.00
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            256K
L3 cache:            6144K
NUMA node0 CPU(s):   0-3
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single pti tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp
```

but our target machine is not much different in specs.

**Comment by jraygauthier** [comment]
> For the other error messages we would like to know,

I'm curious about what are the errors you are referring to?

**Comment by TIS-Arne** [comment]
@jraygauthier 

We are unable to reproduce this issue. Could you please send a small script that triggers the issue for you?

**Comment by jraygauthier** [comment]
@TIS-Arne: Hi, I am on vacation this week, back the next. Leave me some time to create a sample script to reproduce. Indeed it might be that recent changes @TIS-Stefan pushed to `master` in response to this issue fix the problem (TBD). I shall do likewise for #287.

**Comment by jraygauthier** [comment]
@TIS-Stefan :

Applying the following changsets on top of V0.11.1: 

 -  `28d8190b10e342da87a8e808f73130ad92fd4f87`: gst: Do not call src_stop when not running
 -  `c38c63b753939c099d398788b8a522f9e8b6a9b6`: v4l2: Fix pipe file descriptor handling


I saw the following critical gstreamer errors when running my unit tests:

```bash 
$ ./launch-unittest.sh | bcat

(pytest:25483): GStreamer-CRITICAL **: 11:24:21.365: Failed to deactivate pad tcamsrc32:src, very bad

(pytest:25483): GStreamer-CRITICAL **: 11:26:22.917: Failed to deactivate pad tcamsrc61:src, very bad

# ... A stream of those here.

(pytest:25483): GStreamer-CRITICAL **: 11:37:43.445: Failed to deactivate pad tcamsrc586:src, very bad

(pytest:25483): GStreamer-CRITICAL **: 11:37:48.038: Failed to deactivate pad tcamsrc593:src, very bad
```

Whenever one of these error occurs, one the test fails with a timeout awaiting on expected camera frames (which were produced when using V0.9.1).I 

It however does not seem to hang anymore.

Running with debug, I no longer can see the `Error closing udev monitoring pipe` it now outputs the following errors:

```bash
44:34.555 DEBUG   tiscamera_pipeline.py              416  TISCameraPipeline.start: set_state(Gst.State.PLAYING)
44:34.556 INFO    contextlib.py                      119  TISCameraPipeline.start - end
2343867    <INFO> /build/source/src/ImageSource.cpp:68: Source changed to state STOPPED
2343891    <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:567: Stopped stream
2343919    <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:171: Requested format change to 'format=RGGB,width=2448,height=2048,framerate=35.000000' 42474752
2343938    <ERROR> /build/source/src/v4l2/V4l2Device.cpp:188: Error while calling VIDIOC_REQBUFS to empty buffer queue. Device or resource busy
0:00:07.939177595 29250      0x22ff850 ERROR                tcamsrc gsttcamsrc.cpp:1106:send_log_to_bus: Backend reported error: /build/source/src/v4l2/V4l2Device.cpp:188: Error while calling VIDIOC_REQBUFS to empty buffer queue.  
2344007    <ERROR> /build/source/src/v4l2/V4l2Device.cpp:221: Error while setting format 'Device or resource busy'
0:00:07.939234727 29250      0x22ff850 ERROR                tcamsrc gsttcamsrc.cpp:1106:send_log_to_bus: Backend reported error: /build/source/src/v4l2/V4l2Device.cpp:221: Error while setting format ' '
0:00:07.939250658 29250      0x22ff850 ERROR                tcamsrc gsttcamsrc.cpp:977:gst_tcam_src_set_caps: Unable to set format in device
0:00:07.939275670 29250      0x22ff850 WARN                 basesrc gstbasesrc.c:3055:gst_base_src_loop:<tcamsrc24> error: Internal data stream error.
0:00:07.939287087 29250      0x22ff850 WARN                 basesrc gstbasesrc.c:3055:gst_base_src_loop:<tcamsrc24> error: streaming stopped, reason not-negotiated (-4)

# Some shutdown code on timeout.

44:45.931 INFO    contextlib.py                      112  TISCameraPipeline.stop - begin
44:45.931 DEBUG   tiscamera_pipeline.py              456  TISCameraPipeline.stop: set_state(Gst.State.PAUSED)
44:45.931 DEBUG   tiscamera_pipeline.py              458  TISCameraPipeline.stop: set_state(Gst.State.NULL)

(pytest:29250): GStreamer-CRITICAL **: 11:44:45.932: Failed to deactivate pad tcamsrc59:src, very bad

# ...

44:36.727 DEBUG   tiscamera_pipeline.py              416  TISCameraPipeline.start: set_state(Gst.State.PLAYING)
2596448    <INFO> /build/source/src/ImageSource.cpp:68: Source changed to state STOPPED
2596593    <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:567: Stopped stream
44:36.728 INFO    contextlib.py                      119  TISCameraPipeline.start - end
2596621    <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:171: Requested format change to 'format=RGGB,width=2448,height=2048,framerate=35.000000' 42474752
2596641    <ERROR> /build/source/src/v4l2/V4l2Device.cpp:188: Error while calling VIDIOC_REQBUFS to empty buffer queue. Device or resource busy
0:00:10.110474214 29250      0x22ff850 ERROR                tcamsrc gsttcamsrc.cpp:1106:send_log_to_bus: Backend reported error: /build/source/src/v4l2/V4l2Device.cpp:188: Error while calling VIDIOC_REQBUFS to empty buffer queue.  
2596707    <ERROR> /build/source/src/v4l2/V4l2Device.cpp:221: Error while setting format 'Device or resource busy'
0:00:10.110529072 29250      0x22ff850 ERROR                tcamsrc gsttcamsrc.cpp:1106:send_log_to_bus: Backend reported error: /build/source/src/v4l2/V4l2Device.cpp:221: Error while setting format ' '
0:00:10.110542985 29250      0x22ff850 ERROR                tcamsrc gsttcamsrc.cpp:977:gst_tcam_src_set_caps: Unable to set format in device
0:00:10.110567603 29250      0x22ff850 WARN                 basesrc gstbasesrc.c:3055:gst_base_src_loop:<tcamsrc31> error: Internal data stream error.
0:00:10.110579875 29250      0x22ff850 WARN                 basesrc gstbasesrc.c:3055:gst_base_src_loop:<tcamsrc31> error: streaming stopped, reason not-negotiated (-4)
44:36.731 DEBUG   mcu_acquisition_device_async.py    141  MCUAcquisitionDeviceAsync: start_acquisition async begin

# Some shutdown code on timeout.

44:45.931 INFO    contextlib.py                      112  TISCameraPipeline.stop - begin
44:45.931 DEBUG   tiscamera_pipeline.py              456  TISCameraPipeline.stop: set_state(Gst.State.PAUSED)
44:45.931 DEBUG   tiscamera_pipeline.py              458  TISCameraPipeline.stop: set_state(Gst.State.NULL)

(pytest:29250): GStreamer-CRITICAL **: 11:44:45.932: Failed to deactivate pad tcamsrc59:src, very bad

# ...
```

**Comment by TIS-Edgar** [comment]
The pad error has been fixed and has already been pushed to master. Please let us know if this fixes your issues.

**Comment by jraygauthier** [comment]
Hi @TIS-Edgar, I do not have access to my whole ext trigger kit today and limited to soft triggers. Using the latest, it seems to be better using only soft trigs runnin on top of kernel 4.19 and with my patch to comment out the problematic log. I will have to validate running all my automated tests new week.

I think I might have been bitten by either #296 or #295.

**Comment by jraygauthier** [comment]
Hi again @TIS-Edgar . I can still reproduce systematically using b454a8bb06dbcb9e3d3434a03a72d3d2fa44c11e even only with software trigs on kernel 4.19.

See "Update 2019/07/02" section in the description for the script used to reproduce.

**Comment by TIS-Edgar** [comment]
Hi there,
To avoid confusion. 
The pipeline does not hang but receives segfaults. Is that correct?
Is there a reason you are manually deleting the pipeline and source?

It is enough to simply set both to `None`. set_state(Gst.State.NULL) will handle almost all of the required cleanup.

For comparison:
When you compile tiscamera with `BUILD_TESTS=ON` your build directory will contain a test in `tests/integration/start_stop`. 
When you execute `start-stop-runner-sh --caps <caps> --serial` it will repeatedly call an executable, run for ~5 seconds and stop without having problems.

So this segfault has something to do with the way your script is handling the building/destruction of the pipeline/src.  How that can cause this I still do not know.
Is this a blocking issue for you or can you work around it until it is fixed?

**Comment by jraygauthier** [comment]
@TIS-Edgar : 

 > The pipeline does not hang but receives segfaults. Is that correct?

For this issue, the pipeline really hangs after multiple iterations and does not segfault (on top of master, with #295 patch and on kernel 4.19). The segfault is the separate #295 issue. I have another separate issue for when used on top of kernel 4.14 #296 which leads to yet another kind of error.

Note however that on my home machine, I had no success in reproducing with this script. It might be some deadlock or timing issue which is dependent on the particular hw.

 >  Is there a reason you are manually deleting the pipeline and source?

To reproduce what is occurring in the actual application, when running unit tests, etc. The pipeline and the  source are not kept alive forever but only during the acquisition process.

This issue **is the only one for which we have no workaround** (but to revert to using 0.9.1). The hang prevent us from running or tests. 

In the upcoming days I will have much less time to work on these issues though as we're releasing and I'm working on another aspect of the application / environment. I should come back to this at some point later.

**Comment by jraygauthier** [comment]
@TIS-Edgar 

## Reproducing the reentrancy issue with more logs

[tiscamera-unstable-patched-logs]: https://github.com/jraygauthier/tiscamera/commit/5ab0e793ff88a97f09e7728455a2e6c3026d704e

Running with the [tiscamera-unstable-patched-logs] version.


### Non hanging case

```log
03:36.833 DEBUG   tiscamera_pipeline.py              504  TISCameraPipeline.stop: set_state(Gst.State.PAUSED)
03:36.833 DEBUG   tiscamera_pipeline.py              506  TISCameraPipeline.stop: set_state(Gst.State.NULL)
0:00:05.798763228 31741      0x1c56d50 WARN                 tcamsrc gsttcamsrc.cpp:1252:gst_tcam_src_stop: gst_tcam_src_stop: init
0:00:05.798820033 31741      0x1c56d50 WARN                 tcamsrc gsttcamsrc.cpp:1257:gst_tcam_src_stop: gst_tcam_src_stop -> notify_all
0:00:05.798837518 31741      0x1c56d50 WARN                 tcamsrc gsttcamsrc.cpp:1262:gst_tcam_src_stop: gst_tcam_src_stop -> unique_lock
0:00:05.798850894 31741      0x1c56d50 WARN                 tcamsrc gsttcamsrc.cpp:1264:gst_tcam_src_stop: gst_tcam_src_stop -> stop_stream
2251480    <INFO> /build/source/src/ImageSource.cpp:68: Source changed to state STOPPED
2251943    <INFO> /build/source/src/v4l2/V4l2Device.cpp:585: notifying callbacks about lost device
0:00:05.804453533 31741      0x1ce7d20 WARN                 tcamsrc gsttcamsrc.cpp:1050:gst_tcam_src_device_lost_callback: gst_tcam_src_device_lost_callback: init
0:00:05.804594361 31741      0x1ce7d20 WARN                 tcamsrc gsttcamsrc.cpp:1057:gst_tcam_src_device_lost_callback:<tcamsrc9> error: Device lost (25910066)
0:00:05.804657283 31741      0x1ce7d20 WARN                 tcamsrc gsttcamsrc.cpp:1065:gst_tcam_src_device_lost_callback:<tcamsrc9> error: Device lost
0:00:05.804693342 31741      0x1ce7d20 WARN                 tcamsrc gsttcamsrc.cpp:1069:gst_tcam_src_device_lost_callback: gst_tcam_src_device_lost_callback -> gst_tcam_src_stop
0:00:05.804707804 31741      0x1ce7d20 WARN                 tcamsrc gsttcamsrc.cpp:1252:gst_tcam_src_stop: gst_tcam_src_stop: init
0:00:05.804720392 31741      0x1ce7d20 WARN                 tcamsrc gsttcamsrc.cpp:1257:gst_tcam_src_stop: gst_tcam_src_stop -> notify_all
0:00:05.804738695 31741      0x1ce7d20 WARN                 tcamsrc gsttcamsrc.cpp:1262:gst_tcam_src_stop: gst_tcam_src_stop -> unique_lock
2252357    <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:567: Stopped stream
2252393    <INFO> /build/source/src/ImageSink.cpp:55: Pipeline stopped playing
0:00:05.806360619 31741      0x1c56d50 WARN                 tcamsrc gsttcamsrc.cpp:1266:gst_tcam_src_stop: gst_tcam_src_stop -> unlock
0:00:05.806400496 31741      0x1ce7d20 WARN                 tcamsrc gsttcamsrc.cpp:1264:gst_tcam_src_stop: gst_tcam_src_stop -> stop_stream
0:00:05.806416961 31741      0x1ce7d20 WARN                 tcamsrc gsttcamsrc.cpp:1266:gst_tcam_src_stop: gst_tcam_src_stop -> unlock
0:00:05.806429407 31741      0x1ce7d20 WARN                 tcamsrc gsttcamsrc.cpp:1269:gst_tcam_src_stop: gst_tcam_src_stop -> gst_element_send_event
0:00:05.806714122 31741      0x1c56d50 WARN                 tcamsrc gsttcamsrc.cpp:1269:gst_tcam_src_stop: gst_tcam_src_stop -> gst_element_send_event
0:00:05.806748030 31741      0x1c56d50 WARN                 tcamsrc gsttcamsrc.cpp:1273:gst_tcam_src_stop: gst_tcam_src_stop: end
0:00:05.807435647 31741      0x1ce7d20 WARN                 tcamsrc gsttcamsrc.cpp:1273:gst_tcam_src_stop: gst_tcam_src_stop: end
0:00:05.807484769 31741      0x1ce7d20 WARN                 tcamsrc gsttcamsrc.cpp:1071:gst_tcam_src_device_lost_callback: gst_tcam_src_device_lost_callback: end
2264005    <INFO> /build/source/src/CaptureDeviceImpl.cpp:159: Closed device DFM 37UX264-ML.
# Everything alright here.
```

### Hanging case

```log
03:44.888 DEBUG   tiscamera_pipeline.py              504  TISCameraPipeline.stop: set_state(Gst.State.PAUSED)
03:44.889 DEBUG   tiscamera_pipeline.py              506  TISCameraPipeline.stop: set_state(Gst.State.NULL)
0:00:05.703139131 31789      0x1247b50 WARN                 tcamsrc gsttcamsrc.cpp:1252:gst_tcam_src_stop: gst_tcam_src_stop: init
0:00:05.703188087 31789      0x1247b50 WARN                 tcamsrc gsttcamsrc.cpp:1257:gst_tcam_src_stop: gst_tcam_src_stop -> notify_all
0:00:05.703204712 31789      0x1247b50 WARN                 tcamsrc gsttcamsrc.cpp:1262:gst_tcam_src_stop: gst_tcam_src_stop -> unique_lock
0:00:05.703218308 31789      0x1247b50 WARN                 tcamsrc gsttcamsrc.cpp:1264:gst_tcam_src_stop: gst_tcam_src_stop -> stop_stream
2336395    <INFO> /build/source/src/ImageSource.cpp:68: Source changed to state STOPPED
2336822    <INFO> /build/source/src/v4l2/V4l2Device.cpp:585: notifying callbacks about lost device
2336891    <DEBUG> /build/source/src/v4l2/V4l2Device.cpp:567: Stopped stream
0:00:05.704613046 31789      0x12da120 WARN                 tcamsrc gsttcamsrc.cpp:1050:gst_tcam_src_device_lost_callback: gst_tcam_src_device_lost_callback: init
2336935    <INFO> /build/source/src/ImageSink.cpp:55: Pipeline stopped playing
0:00:05.704712388 31789      0x1247b50 WARN                 tcamsrc gsttcamsrc.cpp:1266:gst_tcam_src_stop: gst_tcam_src_stop -> unlock
0:00:05.704735024 31789      0x1247b50 WARN                 tcamsrc gsttcamsrc.cpp:1269:gst_tcam_src_stop: gst_tcam_src_stop -> gst_element_send_event
0:00:05.704758802 31789      0x1247b50 WARN                 tcamsrc gsttcamsrc.cpp:1273:gst_tcam_src_stop: gst_tcam_src_stop: end
0:00:05.704765318 31789      0x12da120 WARN                 tcamsrc gsttcamsrc.cpp:1057:gst_tcam_src_device_lost_callback:<tcamsrc9> error: Device lost (25910066)
0:00:05.704836545 31789      0x12da120 WARN                 tcamsrc gsttcamsrc.cpp:1065:gst_tcam_src_device_lost_callback:<tcamsrc9> error: Device lost
0:00:05.704872546 31789      0x12da120 WARN                 tcamsrc gsttcamsrc.cpp:1069:gst_tcam_src_device_lost_callback: gst_tcam_src_device_lost_callback -> gst_tcam_src_stop
0:00:05.704885883 31789      0x12da120 WARN                 tcamsrc gsttcamsrc.cpp:1252:gst_tcam_src_stop: gst_tcam_src_stop: init
0:00:05.704897681 31789      0x12da120 WARN                 tcamsrc gsttcamsrc.cpp:1257:gst_tcam_src_stop: gst_tcam_src_stop -> notify_all
0:00:05.704915839 31789      0x12da120 WARN                 tcamsrc gsttcamsrc.cpp:1262:gst_tcam_src_stop: gst_tcam_src_stop -> unique_lock
0:00:05.704926563 31789      0x12da120 WARN                 tcamsrc gsttcamsrc.cpp:1264:gst_tcam_src_stop: gst_tcam_src_stop -> stop_stream
0:00:05.704938721 31789      0x12da120 WARN                 tcamsrc gsttcamsrc.cpp:1266:gst_tcam_src_stop: gst_tcam_src_stop -> unlock
0:00:05.704948973 31789      0x12da120 WARN                 tcamsrc gsttcamsrc.cpp:1269:gst_tcam_src_stop: gst_tcam_src_stop -> gst_element_send_event
# .. Hung there, unresponsive.
```


[tiscamera-unstable-patched-workaround]: https://github.com/jraygauthier/tiscamera/commit/ffc6fd777fef968f2d28b65621b1110214b9fdd6

What fixes the problem on our side is the use the
[tiscamera-unstable-patched-workaround] patched version which simply comment out the
`gst_tcam_src_stop` call from `gst_tcam_src_device_lost_callback`.



[tiscamera-unstable-patched-recmut]: https://github.com/jraygauthier/tiscamera/commit/55355609337674365244cc26ad76765b1baddb03

We also tested another patched version [tiscamera-unstable-patched-recmut] which
replace the `std::mutex` by a `std::recursive_mutex`. However with this version
we surprisingly still observe the hanging issue. We attempted this as we **known** that
the current **reentrant** use of `std::mutex` should result in **undefined behaviour**. Some refs:

 -  [std::mutex - cppreference.com](https://en.cppreference.com/w/cpp/thread/mutex):

     >  mutex offers exclusive, **non-recursive** ownership semantics

     >  A calling thread **must not** own the mutex prior to calling lock or try_lock.

 -  [std::recursive_mutex - cppreference.com](https://en.cppreference.com/w/cpp/thread/recursive_mutex)

     >  recursive_mutex offers exclusive, **recursive ownership** semantics: 

     >  A calling thread owns a recursive_mutex for a period of time that starts
     >  when it successfully calls either lock or try_lock. During this period,
     >  the thread **may make additional calls to lock or try_lock**(. The period of
     >  ownership ends when the thread makes a matching number of calls to
     >  unlock.


As patch [tiscamera-unstable-patched-workaround] is satisfactory for our use
case (i.e: the device remains connected at all times), we most likely *won't be
spending much more hours* on this.

However, I **strongly** suggest that this rentrancy issue be fixed *as soon as possible*
as it will be a pain for others to deal with it.


Some potential quick fixes (i.e: removing the possibility of reentrant calls to
`gst_tcam_src_stop`):

 -  Detect when `gst_tcam_src_device_lost_callback` comes from the
    `gst_tcam_src_stop` context and avoid calling `gst_tcam_src_stop` again when
    this is the case.


Some questions to be answered if reentrancy is really meant to be a property of the `gst_tcam_src_stop` function:

 -  As mentionned above: shoud `std::recursive_mutex` be used instead `std::mutex`?

 -  Unprotected `gst_element_send_event(GST_ELEMENT(self), gst_event_new_eos());`?

    Won't it be called twice in the reentrant case. Could it be an issue?

 -  Potential double finalization: `gst_tcam_src_finalize`?

 -  Use of condition variable `self->cv`?

**Comment by TIS-Edgar** [comment]
Hi,

That is a lot of work you put into that.
I will answer your questions to the best of my knowledge.

- should std::recursive_mutex be used instead std::mutex?
No. A recursive_mutex has the purpose of allowing the same thread that already holds the lock to lock the mutex multiple times. Since the call to `gst_tcam_src_stop` from the device_lost_callback is not recursive in any way this will not change anything. The EOS event is handled in a different thread thus the normal mutex suffices. 

- Unprotected gst_element_send_event(GST_ELEMENT(self), gst_event_new_eos());
I assume with unprotected you mean not locked by the mutex? The end-of-stream event if handled by the event loop one of the central parts of gstreamer. 
Once such an event is received all polling the source will stop i.e. gst_tcam_src_create will not be called any more and all elements down the pipeline will flush their data and transmit the eos event.
Sending multiple such events should make no difference as they are handled in a different thread context and all elements handle such an event exactly once.

- Potential double finalization: gst_tcam_src_finalize?
finalize is called when the plugin is destroyed. not the instance of the tcamsrc element.


The condition_variable self->cv  is used to enforce a waiting position in `gst_tcam_src_create`.
This function has the task to fill gstbuffer with image data once an image is ready.
This image 'ready' state is signaled in `gst_tcam_src_sh_callback`.
The reason for this is that the capture thread (V4l2Device.cpp::work_thread ->sh_callback) is independent from the gstreamer thread (create). Both are independent of the device_lost thread as that is simply for the case of "something went wrong, stop the stream".

Instead of de facto disabling device_lost may I suggest the following:

```
    self->is_running = FALSE;
    gst_element_send_event(GST_ELEMENT(self), gst_event_new_eos());
    // gst_tcam_src_stop(GST_BASE_SRC(self));
```
This will cause the pipeline itself to call gst_tcam_src_stop and is likely the better implementation in comparison to the current state.

One bug we recently discovered prevents a correct shutdown and _may_ be related to your issue.
[This patch](https://gist.github.com/TIS-Edgar/ebd6f5039d119909b052c85b55e180c8) addresses both. I'd understand if at this point you simply prefer to keep you solution and be done with it.

**Comment by jraygauthier** [comment]
> Since the call to gst_tcam_src_stop from the device_lost_callback is not recursive in any way this will not change anything. The EOS event is handled in a different thread thus the normal mutex suffices.

I just noticed that indeed the device lost event is called from a totally separate thread. It does indeed makes the use of `std::recursive_mutex` unnecessary. 

> I assume with unprotected you mean not locked by the mutex?

Yes I was.

> Sending multiple such events should make no difference as they are handled in a different thread context and all elements handle such an event exactly once.

Alright. That should make it fine then.

> finalize is called when the plugin is destroyed

Alright, so unrelated.

> The condition_variable self->cv is used to enforce a waiting position in gst_tcam_src_create

Alright too.

However, with the assumption of the above being correct, this still does not explains why calling `gst_tcam_src_stop` directly from the `device_lost_callback` would hang. There must be something else that is wrong and should be understood.

I tested the [patch](cb133de28d978616659b04467a3f425c611a7025) you provided and it seem to fix the hang too.

Some more things I noticed by looking at the source code in an attempt to make sure I understood the various above answers: 

 -  `V4l2Device::is_stream_on` should at least be atomic as it is mutated from multiple threads and read from even more threads.

 -  Same thing applies to `_GstTcamSrc::is_running` which now can be set from the device lost thread too.

 -  It is really hard to understand how each states in the various instances (`V4l2Device`, `_GstTcamSrc`) are shared by the various threads and what kind of model has been taken to ensure thread-safety. For example, the various atomic variables in `V4l2Device` 're access has been made atomic individually. Is is sufficient under some assumption or should there been atomicity as a whole? 

    Some comments over each of these states / state groups would greatly helper a reader understand how developer ensured to preserve correctness of the program.

**Comment by TIS-Edgar** [comment]
You are right with wanting to set a few more flags to atomic. I will apply a few changes to the development branch to address that.

The documentation about internal setups is indeed lacking.
I will add some comments as well as some descriptions to the documentation to make things more 'accessible'.

All programs basically have the following threads:

- the device index thread: started when tcambin/src query the available devices. once a device is open. It is also used for device-lost events.
- Your main thread: starts the stream by telling gstreamer to go into PLAYING
- the gstreamer main stream: puts tcamsrc into PLAYING
- the capture thread: tcamsrc starts this one. It waits until the backend (v4l2, aravis, libusb) says an image is ready and passes it (together with meta information like capture time) to a callback. This notifies gst_tcam_src_create about an image.
- the 'device lost' thread: started by tcamsrc to get notifications if the device does not respond/ is lost
- the streaming thread: this one calls gst_tcam_src_create waiting for new images to iterate over all used elements until the end of the pipeline or a queue is reached. If your streaming thread is identical to your display thread you may encounter problems.

All additional threads are either started in another context by the application and have nothing to do with the image capture or are started by a single gstreamer element for faster image processing. tcamwhitebalance is such an example as it launches work threads to evenly distribute the application of whitebalance values onto the image across all cores.

**Comment by jraygauthier** [comment]
Thanks for this overview, will keep a reference to it.

Is the patch fixing this issue already part of the development branch? Can you provide a ref to the dev branch changeset?

If so (both of the above), I think the issue can be closed.

**Comment by TIS-Edgar** [comment]
Is the patch fixing this issue already part of the development branch?

It is now!
Relevant commits are:  
c3e3d26d4595dc6861440221010660af02987cc4
628b6ff55885e31fe381e52712d5c2e08f5a961f

For an overview over the changes affecting the development branch, please refer to the [CHANGELOG.md](https://github.com/TheImagingSource/tiscamera/blob/development/CHANGELOG.md) of the development branch.
If there are any open questions concerning this issue, lacking documentation or the changelog, please let me know.

**Comment by jraygauthier** [comment]
Excellent. Thanks for the prompt support @TIS-Edgar. This issue has been resolved for us. Closing.
